diff --git a/misc/states.js b/misc/states.js
index 6d98da8..1218169 100644
--- a/misc/states.js
+++ b/misc/states.js
@@ -177,7 +177,8 @@ states.Dependent.prototype = {
 
       // By adding "trigger: true", we ensure that state changes don't go into
       // infinite loops.
-      this.element.trigger({ type: 'state:' + this.state, value: value, trigger: true });
+      this.element.triggerHandler({ type: 'state:' + this.state, value: value, trigger: true });
+      $(document).trigger({ type: 'state:' + this.state, value: value, trigger: true, states_target: this.element.get(0) });
     }
   },
 
@@ -342,14 +343,16 @@ states.Trigger.prototype = {
       var value = valueFn.call(this.element, e);
       // Only trigger the event if the value has actually changed.
       if (oldValue !== value) {
-        this.element.trigger({ type: 'state:' + this.state, value: value, oldValue: oldValue });
+        this.element.triggerHandler({ type: 'state:' + this.state, value: value, oldValue: oldValue });
+        $(document).trigger({ type: 'state:' + this.state, value: value, oldValue: oldValue, states_target: this.element.get(0) });
         oldValue = value;
       }
     }, this));
 
     states.postponed.push($.proxy(function () {
       // Trigger the event once for initialization purposes.
-      this.element.trigger({ type: 'state:' + this.state, value: oldValue, oldValue: null });
+      this.element.triggerHandler({ type: 'state:' + this.state, value: oldValue, oldValue: null });
+      $(document).trigger({ type: 'state:' + this.state, value: oldValue, oldValue: null, states_target: this.element.get(0) });
     }, this));
   }
 };
@@ -480,6 +483,7 @@ $(document).bind('state:disabled', function(e) {
   // Only act when this change was triggered by a dependency and not by the
   // element monitoring itself.
   if (e.trigger) {
+    e.target = e.states_target || e.target;
     $(e.target)
       .attr('disabled', e.value)
         .closest('.form-item, .form-submit, .form-wrapper').toggleClass('form-disabled', e.value)
@@ -492,6 +496,7 @@ $(document).bind('state:disabled', function(e) {
 
 $(document).bind('state:required', function(e) {
   if (e.trigger) {
+    e.target = e.states_target || e.target;
     if (e.value) {
       $(e.target).closest('.form-item, .form-wrapper').find('label').append('<span class="form-required">*</span>');
     }
@@ -503,18 +508,21 @@ $(document).bind('state:required', function(e) {
 
 $(document).bind('state:visible', function(e) {
   if (e.trigger) {
+      e.target = e.states_target || e.target;
       $(e.target).closest('.form-item, .form-submit, .form-wrapper').toggle(e.value);
   }
 });
 
 $(document).bind('state:checked', function(e) {
   if (e.trigger) {
+    e.target = e.states_target || e.target;
     $(e.target).attr('checked', e.value);
   }
 });
 
 $(document).bind('state:collapsed', function(e) {
   if (e.trigger) {
+    e.target = e.states_target || e.target;
     if ($(e.target).is('.collapsed') !== e.value) {
       $('> legend a', e.target).click();
     }
